File: backend/app/services/gb45438.py
import re
from html import unescape

from app.models import EvidenceSignal

TC260_AIGC_NAMESPACE = "http://www.tc260.org.cn/ns/AIGC/1.0/"
TC260_FIELDS = [
    "Label",
    "ContentProducer",
    "ProduceID",
    "ReservedCode1",
    "ContentPropagator",
    "PropagateID",
    "ReservedCode2",
]
SIGNATURE_TERMS = [
    b'"AIGC"',
    b'"aigc"',
    b'"aigc_info"',
    b"GB 45438",
    b"GB45438",
    b"AI_GENERATED",
    TC260_AIGC_NAMESPACE.encode("utf-8"),
]


def inspect_gb45438(image_bytes: bytes) -> EvidenceSignal:
    text = image_bytes.decode("utf-8", errors="ignore")
    matched_terms = [term.decode("utf-8", errors="ignore") for term in SIGNATURE_TERMS if term in image_bytes]
    xmp_fields = _extract_tc260_xmp_fields(text)
    detected = bool(matched_terms or xmp_fields)

    return EvidenceSignal(
        detected=detected,
        status="detected" if detected else "absent",
        summary=(
            "Possible GB 45438/AIGC metadata or byte-level marker was found."
            if detected
            else "No GB 45438/AIGC metadata or byte-level marker was found by v0 scan."
        ),
        details={
            "matched_terms": matched_terms,
            "tc260_namespace_detected": TC260_AIGC_NAMESPACE in text,
            "xmp_fields": xmp_fields,
            "scan_mode": "tc260_xmp_and_byte_markers",
        },
    )


def _extract_tc260_xmp_fields(text: str) -> dict[str, str]:
    fields = {}
    for field in TC260_FIELDS:
        value = _extract_xml_value(text, field)
        if value:
            fields[field] = value
    return fields


def _extract_xml_value(text: str, local_name: str) -> str | None:
    # Handles both prefixed tags like <AIGC:Label> and default namespace tags like <Label>.
    pattern = rf"<(?:[A-Za-z0-9_.-]+:)?{re.escape(local_name)}\b[^>]*>(.*?)</(?:[A-Za-z0-9_.-]+:)?{re.escape(local_name)}>"
    match = re.search(pattern, text, flags=re.IGNORECASE | re.DOTALL)
    if not match:
        return None
    return unescape(re.sub(r"\s+", " ", match.group(1)).strip())[:300]

File: extension/analysis/gb45438.js
// GB 45438 / TC260 AIGC marker scan, ported from backend app/services/gb45438.py.

const TC260_AIGC_NAMESPACE = "http://www.tc260.org.cn/ns/AIGC/1.0/";

const TC260_FIELDS = [
  "Label",
  "ContentProducer",
  "ProduceID",
  "ReservedCode1",
  "ContentPropagator",
  "PropagateID",
  "ReservedCode2",
];

const SIGNATURE_TERMS = [
  '"AIGC"',
  '"aigc"',
  '"aigc_info"',
  "GB 45438",
  "GB45438",
  "AI_GENERATED",
  TC260_AIGC_NAMESPACE,
];

const latin1 = new TextDecoder("latin1");
const utf8 = new TextDecoder("utf-8", { fatal: false, ignoreBOM: true });

export function inspectGb45438(bytes) {
  const byteText = latin1.decode(bytes);
  const xmpText = utf8.decode(bytes);

  const matchedTerms = SIGNATURE_TERMS.filter((term) => byteText.includes(term));
  const xmpFields = extractTc260XmpFields(xmpText);
  const detected = matchedTerms.length > 0 || Object.keys(xmpFields).length > 0;

  return {
    checked: true,
    detected,
    status: detected ? "detected" : "absent",
    summary: detected
      ? "Possible GB 45438/AIGC metadata or byte-level marker was found."
      : "No GB 45438/AIGC metadata or byte-level marker was found by v0 scan.",
    details: {
      matched_terms: matchedTerms,
      tc260_namespace_detected: byteText.includes(TC260_AIGC_NAMESPACE),
      xmp_fields: xmpFields,
      scan_mode: "tc260_xmp_and_byte_markers",
    },
  };
}

function extractTc260XmpFields(text) {
  const fields = {};
  for (const field of TC260_FIELDS) {
    const value = extractXmlValue(text, field);
    if (value) {
      fields[field] = value;
    }
  }
  return fields;
}

function extractXmlValue(text, localName) {
  const escaped = localName.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const pattern = new RegExp(
    `<(?:[A-Za-z0-9_.-]+:)?${escaped}\\b[^>]*>([\\s\\S]*?)</(?:[A-Za-z0-9_.-]+:)?${escaped}>`,
    "i",
  );
  const match = pattern.exec(text);
  if (!match) {
    return null;
  }
  const collapsed = match[1].replace(/\s+/g, " ").trim();
  return decodeEntities(collapsed).slice(0, 300);
}

function decodeEntities(value) {
  return value
    .replace(/&#x([0-9a-fA-F]+);/g, (_, hex) => safeCodePoint(parseInt(hex, 16)))
    .replace(/&#(\d+);/g, (_, dec) => safeCodePoint(parseInt(dec, 10)))
    .replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&amp;/g, "&");
}

function safeCodePoint(code) {
  if (!Number.isFinite(code) || code < 0 || code > 0x10ffff) {
    return "";
  }
  try {
    return String.fromCodePoint(code);
  } catch {
    return "";
  }
}
